accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,530 @@
+
+Privately scoped packages, or packages from private repositories, will not be signed or published to the public ledger. This will be determined by interrogating the npm registry at publish time. Users will be able to override this similar to the `cosign --force` command by passing an argument to `npm publish`, e.g. `--force-build-signatures`.
+
+The entry will be stored and retrieved using the shasum of the package tarball. This will make sure we don't find any signatures for a deleted package that has been resurrected with the same name and version.
+
+There will be no way of deleting an entry if this was published unintentionally to the ledger as it's immutable by design to prevent tampering. The practical consequence of publishing to the ledger is that information about your package (name, version), repository (commit, branch, owner) and workflow (name, id) are uploaded to the public ledger. A possible workaround would be to rename the leaked repo, make it private and delete the old one.
+
+##### Detailed steps to publish
+Interaction between the npm CLI tool running in CI, the npm registry, and the Sigstore components Fulcio and Rekor:
+
+```mermaid
+  sequenceDiagram
+    participant npm_cli
+    participant ci_cd
+    participant fulcio
+    participant rekor
+    participant npm_registry
+    npm_cli->>ci_cd: Request workflow identity (OIDC ID token)
+    npm_cli->>fulcio: Exchange ID token for signing certificate
+    fulcio->>npm_cli: Signing certificate
+    npm_cli->>rekor: Upload signed build provenance attestation
+    rekor->>npm_cli: Rekor entry
+    npm_cli->>npm_registry: Publish package and upload Rekor entry
+    npm_registry->>fulcio: Request signing certificate
+    fulcio->>npm_registry: Signing certificate
+    npm_registry->>rekor: Upload signed release attestations
+    rekor->>npm_registry: Rekor entry
+    npm_registry->>npm_registry: Attach rekor entry to published package
+    npm_registry->>npm_cli: Success
+```
+
+1. `npm publish --with=build-signatures` detects execution within a supported CI/CD environment and initializes token exchange to get a JWT id token from the CI/CD system's OIDC identity provider.
+2. npm CLI creates a new ephemeral key pair, requests a certificate from Fulcio (CA) and authenticates via the JWT id token.
+3. A SLSA build provenance attestation is created based on information available within the JWT and CI system and signs it with the ephemeral key.
+4. npm CLI uploads the completed attestation record on Rekor for public packages.
+5. npm registry creates a release attestation, stating that the version has been accepted and published, and places the record on Rekor for public packages.
+
+#### Verifying
+The package integrity captured by the provenance attestation will be verified by the npm CLI. To begin with, the package integrity should be verified when running the  `npm audit signatures` command and eventually transparently integrated into `npm install` and enabled by default.
```

Comment:
It's not clear to me what security value this provides, absent a policy that requires builds to come from a specific source repo. IIUC, this says: "if there is provenance, it has to come from one of the trusted builders, but it can be built from any source." I could see value if that provenance is then displayed to the user or otherwise acted upon, but IIUC it is not. I don't think this will stop any attacks because the attacker can just built "verifiably" from a fork, then delete the fork.

Could you explain the purpose of this check?

(To be clear, I see the value in verifying on the NPM website, and in laying the groundwork for future policy checks. I was just wondering what the point is in the check as written here.)

Reply:
This is a good point. Just verifying these signatures doesn't say if the provenance has changed but it does verify the package hasn't been tampered with since build. What we can do is start enforcing the presence of provenance information (making it sticky) once it's been set on a package and possibly prevent a publish if this changes, unless you unlink it from your npm dashboard using 2FA for example.

It would also be great to figure out how we could bubble up changes to consumers somehow through the CLI when running these commands. One idea would be to warn on any changes when doing package upgrades in the CLI as we could keep track of before and after. I also envision Dependabot PRs highlighting these changes.

Reply:
Yeah, displaying a warning on provenance change during upgrades would be a great next step. Maybe add that brief explanation and idea to the RFC?

(A subsequent step could be requiring explicit authorization for provenance changes, such as by recording the source repo in the package.lock file. In my opinion that's a decent idea for policy.)

Reply:
Don't want to get too deep into the weeds, but things on my end we've done in the past is have a separate source attestation, or trusted signature against the source. Then the policy becomes both a verification that there is an attestation against the build and then looking in the SLSA provenance for `materials`.

This makes some assumptions though that the builder can't lie about the source it came from. So we need to make sure the GHA flow doesn't allow for the package maintainer piece to lie about the code.